File: delay_cat.py
#!/usr/bin/env python3

import argparse

parser = argparse.ArgumentParser(add_help=False)
parser.add_argument('-d', '--delay', type=float, default=0.5)
args, _ = parser.parse_known_args()

DELAY = args.delay

#----------------
import sys, time, heapq, select

# DELAY = 0.5
pq = []  # (emit_time, line)

stdin = sys.stdin
stdout = sys.stdout

while True:
    now = time.monotonic()

    # 计算下一个需要输出的时间
    timeout = None
    if pq:
        timeout = max(0, pq[0][0] - now)

    # 等待 stdin 可读 或 到达下一个 emit 时间
    r, _, _ = select.select([stdin], [], [], timeout)

    if r:
        line = stdin.readline()
        if not line:   # EOF
            break
        heapq.heappush(pq, (time.monotonic() + DELAY, line))

    # 输出所有到期的行
    now = time.monotonic()
    while pq and pq[0][0] <= now:
        _, line = heapq.heappop(pq)
        stdout.write(line)
        stdout.flush()

# stdin 结束后，flush 剩余
while pq:
    t, line = heapq.heappop(pq)
    time.sleep(max(0, t - time.monotonic()))
    stdout.write(line)
    stdout.flush()

File: delay_cat.pl
#!/usr/bin/env perl

use strict;
use warnings;
use Time::HiRes qw(time sleep);
use Getopt::Long;

my $delay = 0.5;
GetOptions(
    'd|delay=f' => \$delay,
) or die "usage: late_cat [-d|--delay seconds]\n";

$| = 1;  # unbuffer stdout

my @pq;  # priority queue: [emit_time, line]

sub pq_push {
    my ($t, $line) = @_;
    push @pq, [$t, $line];
    @pq = sort { $a->[0] <=> $b->[0] } @pq;
}

sub pq_pop_ready {
    my ($now) = @_;
    my @out;
    while (@pq && $pq[0]->[0] <= $now) {
        push @out, shift @pq;
    }
    return @out;
}

while (1) {
    my $now = time();

    my $timeout;
    if (@pq) {
        $timeout = $pq[0]->[0] - $now;
        $timeout = 0 if $timeout < 0;
    }

    my $rin = '';
    vec($rin, fileno(STDIN), 1) = 1;

    my $n = select($rin, undef, undef, $timeout);

    if ($n) {
        my $line = <STDIN>;
        last unless defined $line;   # EOF
        pq_push(time() + $delay, $line);
    }

    $now = time();
    for my $item (pq_pop_ready($now)) {
        print $item->[1];
    }
}

# flush remaining after EOF
for my $item (@pq) {
    my $wait = $item->[0] - time();
    sleep($wait) if $wait > 0;
    print $item->[1];
}

